enterprise/server/remote_execution/containers/firecracker/firecracker_test.go

```diff
@@ -564,6 +550,356 @@ func TestFirecracker_LocalSnapshotSharing(t *testing.T) {
+	require.NoError(t, err)
+
+	// Write VM-specific data to the log
+	cmd = appendToLog("Fork remote fetch")
+	res = forkedVM2.Exec(ctx, cmd, nil /*=stdio*/)
+	require.NoError(t, res.Error)
+	// The log should contain data written to the original snapshot
+	// and the current VM, but not from any of the other VMs sharing
+	// the same original snapshot
+	require.Equal(t, "Base\nFork remote fetch\n", string(res.Stdout))
+}
+
+// Prints performance data about various Firecracker commands
+func TestFirecracker_RemoteSnapshotSharing_ManualBenchmarking(t *testing.T) {
+	if !*testManualBenchmark {
+		t.Skip()
+	}
+	// Silence the logs so output is easier to read
+	flags.Set(t, "app.log_level", "error")
+	log.Configure()
+
+	rand.Seed(time.Now().UnixNano())
+
+	var containersToCleanup []*firecracker.FirecrackerContainer
+	t.Cleanup(func() {
+		for _, vm := range containersToCleanup {
+			err := vm.Remove(context.Background())
+			assert.NoError(t, err)
+		}
+	})
+
+	var (
+		env        *testenv.TestEnv
+		ctx        context.Context
+		c          *firecracker.FirecrackerContainer
+		task       *repb.ExecutionTask
+		opts       firecracker.ContainerOpts
+		jailerRoot string
+	)
+
+	setup := func() {
+		var err error
+		env = testenv.GetTestEnv(t)
+		flags.Set(t, "executor.enable_local_snapshot_sharing", true)
+		flags.Set(t, "executor.firecracker_enable_nbd", true)
+		flags.Set(t, "executor.firecracker_enable_uffd", true)
+
+		ctx = context.Background()
+		env = getTestEnv(ctx, t)
+		rootDir := testfs.MakeTempDir(t)
+		jailerRoot = tempJailerRoot(t)
+		workDir := testfs.MakeDirAll(t, rootDir, "work")
+
+		env.SetAuthenticator(testauth.NewTestAuthenticator(testauth.TestUsers("US1", "GR1")))
+		filecacheRoot := testfs.MakeDirAll(t, jailerRoot, "filecache")
+		fc, err := filecache.NewFileCache(filecacheRoot, fileCacheSize, false)
+		require.NoError(t, err)
+		fc.WaitForDirectoryScanToComplete()
+		env.SetFileCache(fc)
+
+		task = &repb.ExecutionTask{
+			Command: &repb.Command{
+				// Note: platform must match in order to share snapshots
+				Platform: &repb.Platform{Properties: []*repb.Platform_Property{
+					{Name: "recycle-runner", Value: "true"},
+				}},
+			},
+		}
+		opts = firecracker.ContainerOpts{
+			ContainerImage:         busyboxImage,
+			ActionWorkingDirectory: workDir,
+			VMConfiguration: &fcpb.VMConfiguration{
+				NumCpus:           1,
+				MemSizeMb:         minMemSizeMB, // small to make snapshotting faster.
+				EnableNetworking:  false,
+				ScratchDiskSizeMb: 100,
+			},
+			JailerRoot: jailerRoot,
+		}
+
+		c, err = firecracker.NewContainer(ctx, env, task, opts)
+		require.NoError(t, err)
+		containersToCleanup = append(containersToCleanup, c)
+		err = container.PullImageIfNecessary(ctx, env, c, container.PullCredentials{}, opts.ContainerImage)
+		require.NoError(t, err)
+		err = c.Create(ctx, opts.ActionWorkingDirectory)
+		require.NoError(t, err)
+	}
+
+	// Pausing a new VM
+	for _, enableRemote := range []bool{true, false} {
+		{
```

Comment:
nit: the extra `{` within the loop is unnecessary, since for-loops create their own scope already (here and below)